README.md
# Gittins

This is an R package to calculate Gittins indices for the multi-armed bandit problem.

## Description

This project contains functions written in R to calculate Gittins indices for the Bayesian multi-armed bandit problem with Bernoulli or Normal rewards. Detailed information on the methodology as well as the effect of different computation parameters on speed and accuracy can be found in this [paper](https://arxiv.org/abs/1909.05075) and in my [thesis](http://eprints.lancs.ac.uk/84589/) (in Chapter 5).

## Installation

Install the gittins package from github with:

``` r
# install.packages("devtools")
devtools::install_github("jedwards24/gittins")
```

## Usage

There are two main groups of functions, those for the multi-armed bandit with bernoulli rewards (begin with `bmab_`) and those for the same problem with normal rewards (begin with `nmab_`). 

The functions `bmab_gi()` and `nmab_gi()` will calculate the gittins index of a single arm state, while `bmab_gi_multiple()` and `nmab_gi_multiple()` will calculate indices for the multiple states that will be encountered in a multi-armed bandit problem.

The package also contains functions to calculate other indices, the KGI and GI+, which provide lower and upper bounds 
for the Gittins index.

### Bernoulli Rewards

The following calculates the Gittins index for an arm which has a Bayesian state (prior plus observed) of 
two observations and one success. 

``` r
bmab_gi(1, 2, gamma = 0.9, N = 80, tol = 5e-5)
```

The other arguments used are:  the reward discount factor `gamma` of the problem, the desired accuracy `tol`, and the horizon `N` of the dynamic program. The first of these is a feature of the problem while the last two affect the calculation accuracy and speed. 

The alternative function `bmab_gi_ab()` gives the arm state in alpha/beta values (success/failure counts). 

For multiple states, give the starting state of the problem (here using alpha/beta terminology) and the number of actions that will be taken. These define the possible states the arm could take and therefore the states for which Gittins indices are needed. The following function gives these results in a data frame:

``` r
bmab_gi_multiple(alpha_start = 1, beta_start = 1, num_actions = 10, gamma = 0.9, N = 80, tol = 5e-5)
```

### Normal Rewards

This is similar to the Bernoulli case but with extra parameters. For a a single state:

```r
nmab_gi(0, 1, gamma = 0.9, tau = 1, N = 30, xi = 3, delta = 0.02, tol = 5e-4)
```

The arguments `gamma` and `tau` are problem setting parameters, while the remaining arguments affect 
the solution accuracy and speed. Calculations for the NMAB can be much slower than for the BMAB 
especially if high accuracy is required.

For multiple states, Gittins indices are only needed for a single vector of states given by the parameter `n` since indices for different mean beliefs can then be can be found by a simple transform. 

``` r
nmab_gi_multiple(n_range = 1:10, gamma = 0.9, tau = 1, N = 30, xi = 3, delta = 0.02, tol = 5e-4)
```


NEWS.md
# gittins 0.2.0

Breaking changes:

* `bmab_gi_multiple()` (breaking changes):
  * Now accepts alpha/beta or Sigma/n state inputs.
  * Therefore `bmab_gi_multiple_ab()` is no longer needed and so has been removed.
  * Output is now a data frame with a row for each starting state. The original matrix output is 
  now attached as attributes.
* `nmab_gi_multiple()` give output as a data frame.
* The ordering of some function arguments has been changed.

Other changes:

* Add checks that parameters take appropriate values. 
* The `tol` argument now has a default value (5e-4).
* Code tidying and style changes.
* Documentation edits. 
* Add GPL3 licence.
* Added `@keywords internal` to value functions and `calibrate_arm()` since they will not normally 
be used directly (they are still exported).
* Add package help file with useful links (see `?gittins`).
* Rewrite README.

# gittins 0.1.0

This is the original working version from 2017. It is available on Github under releases/tags.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(gittins)

test_check("gittins")


tests/testthat/test-nmab.R
test_that("nmab_gi gives expected output", {
  x1 <- nmab_gi(0, 1, 0.8, 1, 30, 3, 0.1, 1e-4)
  x2 <- nmab_gi(1, 1, 0.8, 1, 30, 3, 0.1, 1e-4)
  expect_identical(x1 + 1, x2)
  expect_equal(x1, 0.50552724)
})

test_that("nmab_kgi and nmab_giplus give expected output", {
  x3 <- nmab_kgi(0, 1, 0.8, 1, 1e-4)
  x4 <- nmab_giplus(0, 1, 0.8, 1e-4)
  x5 <- nmab_giplus(0, 1, 0.8, 1e-4, upper = TRUE)
  expect_gte(x5, x4)
  expect_equal(x3, 0.44973755)
  expect_equal(x4, 0.63601685)
})



tests/testthat/test-check.R
test_that("`check_numeric()` works.", {
  expect_null(check_numeric(5, "arg"))
  expect_null(check_numeric(5, "arg", 1))
  expect_null(check_numeric(5, "arg", 1, 10))
  expect_error(check_numeric(0, "arg", 1, 10), "`arg` must be greater than 1")
  expect_error(check_numeric(1, "arg", 1, 10), "`arg` must be greater than 1")
  expect_error(check_numeric(5, "arg", 1, 5), "`arg` must be less than 5")
  expect_null(check_numeric(5.5, "arg", 1, 10))
})

test_that("`check_integerish()` works.", {
  expect_null(check_integerish(5, "arg"))
  expect_error(check_integerish(5.5, "arg"), "`arg` must be a whole number")
  expect_null(check_integerish(5, "arg", 5, 9))
  expect_null(check_integerish(5, "arg", 0, 5))
  expect_error(check_integerish(0, "arg", 1, 10), "`arg` must be at least 1")
  expect_error(check_integerish(6, "arg", 1, 5), "`arg` must be no greater than 5")
})


R/bmab.R
#' Function arguments
#'
#' @keywords internal
#' @name bmab_args
#'
#' @param Sigma Numeric > 0. Value of Sigma for the arm.
#' @param n Numeric > Sigma > 0. Value of n for the arm.
#' @param gamma Numeric in (0, 1). Reward discount factor.
#' @param N Integer >= 2. Time horizon used in the calculation..
#' @param tol Numeric > 0. Absolute accuracy required.
NULL

#' Function arguments for value functions
#'
#' @keywords internal
#' @name bmab_v_args
#'
#' @param lambda Reward from the known arm.
#' @param Sigma Value of Sigma for the unknown arm.
#' @param n Value of n for the unknown arm.
NULL

#' Calculate Gittins indices for multiple starting states (Bernoulli rewards)
#'
#' Runs `bmab_gi()` multiple times for a range of arm starting states. The starting states used are
#' determined by the starting state arguments and `num_actions` (see details). Other arguments are
#' passed to `bmab_gi()`. See `?bmab_gi` for argument and calculation details.
#'
#' @details
#' The starting states are either given in alpha/beta or Sigma/n form (strictly one or the other).
#' These are alternative ways of doing the same thing since  `alpha = Sigma` and  `beta = n - Sigma`.
#'
#' The GI are calculated for arms with the following states:
#' * `alpha = alpha_start : (alpha_start_start + num_actions - 1)`,
#' * `beta = beta_start : (beta_start + num_actions - 1)`,
#' * where `alpha + beta <= num_actions + alpha_start + beta_start - 1`.
#'
#' @inheritParams bmab_args
#' @param alpha_start Lowest value of alpha for the arms.
#' @param beta_start Lowest value of beta for the arms.
#' @param Sigma_start Lowest value of Sigma for the arms.
#' @param n_start Lowest value of n for the arms.
#' @param num_actions Determines the number of states GI are calculated for.
#'
#' @return A data frame of starting states and GI values. Extra information is stored in attributes
#' (see examples):
#' * `params`: the parameters used.
#' * `gi_matrix`: the GI values in a matrix (beta x alpha).
#' * `gi_matrix_ns`: the GI values in a matrix (n x Sigma).
#'
#' @examples
#' b1 <- bmab_gi_multiple(1, 1, num_actions = 4, gamma = 0.8, N = 20)
#' b2 <- bmab_gi_multiple(num_actions = 4, gamma = 0.8, N = 20, Sigma_start = 1, n_start = 2)
#' identical(b1, b2) # TRUE
#' # View attributes
#' attr(b1, "params")
#' attr(b1, "gi_matrix")
#' attr(b1, "gi_matrix_ns")
#'
#' @export
bmab_gi_multiple <- function(alpha_start = NULL, beta_start = NULL, num_actions, gamma, N, tol = 5e-4,
                             Sigma_start = NULL, n_start = NULL){
  if (!check_start_args(alpha_start = alpha_start, beta_start = beta_start,
                        Sigma_start = Sigma_start, n_start = n_start)){
    stop("Incorrect number of starting state arguments.\n",
         "You must supply `alpha_start` and `beta_start` OR `Sigma_start` and `n_start`.", call. = FALSE)
  }
  if (!is.null(Sigma_start)){
    if (Sigma_start >= n_start){
      stop("`n_start` must be greater than `Sigma_start`.", call. = FALSE)
    }
    alpha_start <- Sigma_start
    beta_start <- n_start - Sigma_start
  }
  check_numeric(alpha_start, "alpha_start", 0)
  check_numeric(beta_start, "beta_start", 0)
  check_integerish(num_actions, "num_actions", 2L)

  n_states <- as.integer(round(0.5 * num_actions * (num_actions + 1), 0))
  alpha_vec <- beta_vec <- integer(n_states)
  alpha_range <- alpha_start:(alpha_start + num_actions - 1)
  beta_range <- beta_start:(beta_start + num_actions - 1)
  gi_mat <- matrix(rep(NA, num_actions * num_actions), nrow = num_actions, ncol = num_actions,
                   dimnames = list(beta = beta_range, alpha = alpha_range))
  mu <- alpha_start / (alpha_start + beta_range)
  lb_vec <- bmab_kgi(alpha_start, alpha_start + beta_range, gamma)
  cat("Calculating GI values for", n_states, "states\n")
  index <- 1
  pb <- txtProgressBar(min = 0, max = num_actions, style = 3)
  for (a in 1:num_actions){
    betas <- beta_range[1:(num_actions - a + 1)]
    len_betas <- length(betas)
    end_index <- index + len_betas - 1
    alpha_vec[index:end_index] <- rep(alpha_range[a], len_betas)
    beta_vec[index:end_index] <- betas
    index <- end_index + 1
    ub <- 1
    for (b in seq_along(betas)){
      gi_mat[b, a] <- bmab_gi_ab(alpha_range[a], beta_range[b], gamma, N, tol, lb = lb_vec[b], ub = ub)
      ub <- gi_mat[b, a]
    }
    lb_vec <- gi_mat[, a]
    setTxtProgressBar(pb, a)
  }
  close(pb)
  gi_mat_ns <- gi_mat
  for(i in 2:num_actions){
    gi_mat_ns[, i] <- c(rep(NA, i - 1), gi_mat[1:(num_actions - i + 1), i])
  }
  dimnames(gi_mat_ns) <- list(n = beta_start + alpha_range, Sigma = alpha_range)
  n_vec <- alpha_vec + beta_vec
  df <- data.frame(alpha = alpha_vec, beta = beta_vec, gi = as.numeric(na.omit(as.vector(gi_mat))),
                       Sigma = alpha_vec, n = n_vec, stage = n_vec - min(n_vec))
  attr(df, "params") <- list(alpha_start = alpha_start, beta_start = beta_start,
                             Sigma_start = alpha_start, n_start = alpha_start + beta_start,
                             gamma = gamma, N = N, num_actions = num_actions, tol = tol)
  attr(df, "gi_matrix") <- gi_mat
  attr(df, "gi_matrix_ns") <- gi_mat_ns
  df
}

#' Calculate the Gittins index for a single arm (Bernoulli rewards)
#'
#' @description
#' The two versions of this function differ only in their state parameters:
#' * `bmab_gi()` uses `Sigma` and `n` parameters.
#' * `bmab_gi_ab()` uses `alpha` and `beta` parameters
#'
#' These are related by  `alpha = Sigma` (Bayesian number of successes) and  `beta = n - Sigma`
#' (Bayesian number of failures). Then `n` is the Bayesian number of observations. Together
#' with `gamma` (the discount factor for rewards), these define the problem. The remaining arguments
#' are settings for the calculation only (see details).
#'
#' @details
#' The problem has an infinite time horizon, but the dynamic program used to calculate the GI has a finite
#' horizon `N`. For sufficiently large `N`, the calculation can be arbitrarily accurate. In practice,
#' a fairly low value of `N` works well unless `gamma` is close to 1.
#'
#' The `lb` and `ub` arguments can be used to provide a starting interval for calibration if desired.
#' However, for normal use this is not needed as they will be calculated internally if not supplied.
#' So the initial interval is determined as follows:
#' * For lower bound, use `lb` if supplied else use KGI.
#' * For upper bound, use `ub` if supplied else use GI+.
#'
#' @inheritParams bmab_args
#' @param lb Optional lower bound for GI.
#' @param ub Optional upper bound for GI.
#' @seealso For a link to the accompanying paper see [gittins-package].
#'
#' @return A single Gittins index
#' @export
bmab_gi <- function(Sigma, n, gamma, N, tol = 5e-4, lb = NA, ub = NA){
  if (is.na(lb)){
    lb <- bmab_kgi(Sigma, n, gamma)
  }
  if (is.na(ub)){
    ub <- bmab_giplus(Sigma, n, gamma, tol, upper = TRUE)
  }
  check_numeric(Sigma, "Sigma", 0)
  check_numeric(n, "n", 0)
  if (Sigma >= n){
    stop("`n` must be greater than `Sigma`.", call. = FALSE)
  }
  check_numeric(gamma, "gamma", 0, 1)
  check_integerish(N, "N", 2L)
  check_numeric(tol, "tol", 0)
  check_numeric(lb, "lb")
  check_numeric(ub, "ub")
  mean(calibrate_arm(bmab_gi_value, lb, ub, tol, Sigma, n, gamma, N))
}

#' @param alpha Numeric > 0.Value of alpha for the arm.
#' @param beta Numeric > 0. Value of beta for the arm.
#'
#' @rdname bmab_gi
#' @export
bmab_gi_ab <- function(alpha, beta, gamma, N, tol = 5e-4, lb = NA, ub = NA){
  bmab_gi(Sigma = alpha, n = alpha + beta, gamma = gamma, N = N, tol = tol, lb = lb, ub = ub)
}

#' Calculate the GI+ index for a single arm (Bernoulli rewards)
#'
#' The GI+ index is an upper bound for the Gittins index.
#'
#' @inheritParams bmab_args
#' @param upper if TRUE, the upper end of the interval is returned, otherwise the midpoint.
#'
#' @return A GI+ index value.
#' @export
bmab_giplus <- function(Sigma, n, gamma, tol = 5e-4, upper = FALSE){
  interval <- calibrate_arm(bmab_giplus_value, lb = Sigma / n, ub = 1, tol, Sigma, n, gamma)
  if (upper){
    return(interval[2])
  }
  mean(interval)
}

#' Calculate the knowledge gradient index for a single arm (Bernoulli rewards)
#'
#' The KGI is an lower bound for the Gittins index.
#'
#' This is an exact closed form calculation. Arguments `Sigma` and `n` can be supplied as vectors in
#' which case a vector of index values will be returned.
#'
#' @inheritParams bmab_args
#'
#' @return An index value or a vector of values.
#' @export
bmab_kgi <- function(Sigma, n, gamma){
  mu <- Sigma / n
  H <- gamma / (1 - gamma)
  (mu + H * mu * (Sigma + 1) / (n + 1)) / (1 + H * mu)
}

#' Value of one-armed bandit using GI+ (Bernoulli rewards)
#'
#' @inheritParams bmab_v_args
#' @inheritParams bmab_args
#' @keywords internal
#' @return Difference in value between safe and unknown arms.
#' @export
bmab_giplus_value <- function(lambda, Sigma, n, gamma){
  mu <- Sigma / n
  H <- gamma / (1 - gamma)
  value_success <- H * (integrate(continue, lambda, 1, Sigma = Sigma, n = n)[[1]] +
                          lambda * pbeta(lambda, Sigma + 1, n - Sigma))
  value_fail <- H * lambda
  mu + mu * value_success + (1 - mu) * value_fail - lambda / (1 - gamma)
}

#' Helper function only used in bmab_giplus_value()
#' @noRd
continue <- function(x, Sigma, n){
  dbeta(x, Sigma + 1, n - Sigma) * x
}

#' Value of a one-armed bandit (Bernoulli rewards)
#'
#' @inheritParams bmab_v_args
#' @inheritParams bmab_args
#'
#' @keywords internal
#' @return Difference in value between safe and unknown arms.
#' @export
bmab_gi_value <- function(lambda, Sigma, n, gamma, N){
  h <- N + 1
  n_vec <- n:(n + N)
  s_vec <- Sigma:(Sigma + N)
  mu <- outer(s_vec, n_vec, "/")
  value_mat <- matrix(nrow = h, ncol = h)
  # Values of end states
  value_mat[, h] <- pmax(mu[, h], lambda) * gamma^N / (1 - gamma)
  safe_reward <- lambda * gamma^((1:N) - 1) / (1 - gamma)
  # Run DP to get values of other states
  for (i in N:1){
    j <- i + 1
    risky_reward <- mu[1:i, i] * (gamma^(i - 1) + value_mat[2:j, j]) +
      (1 - mu[1:i, i]) * value_mat[1:i, j]
    value_mat[1:i, i] <- pmax(risky_reward, safe_reward[i])
  }
  return(value_mat[1, 1] - lambda / (1 - gamma))
}


R/check.R
#' Check if a vector is a scalar numeric strictly within given bounds
#'
#' Returns `NULL` if okay. Errors otherwise.
#' @param x Object to test.
#' @param name String to use as argument name in error message. Will be enclosed in back ticks.
#' @param min Lower bound for `x`. `x > min`.
#' @param max Upper bound for `x`. `x < max`.
#' @noRd
check_numeric <- function(x, name, min = -Inf, max = Inf, int = FALSE) {
  name <- paste0("`", name, "`")
  if (length(x) > 1 || !is.numeric(x)){
    stop(name, " must be a length 1 numeric.", call. = FALSE)
  }
  if (x <= min){
    stop(name, " must be greater than ", min, ".", call. = FALSE)
  }
  if (x >= max){
    stop(name, " must be less than ", max, ".", call. = FALSE)
  }
  return()
}

#' Check if a vector is a scalar integer like numeric within given bounds
#'
#' Unlike `check_numeric()` this allows equality with bounds.
#' Returns `NULL` if okay. Errors otherwise.
#' @param x Object to test.
#' @param name String to use as argument name in error message. Will be enclosed in back ticks.
#' @param min Lower bound for `x`. `x >= min`.
#' @param max Upper bound for `x`. `x <= max`.
#' @param int Logical. Should `x` be integer like?
#' @noRd
check_integerish <- function(x, name, min = -Inf, max = Inf) {
  name <- paste0("`", name, "`")
  if (length(x) > 1 || !is.numeric(x)){
    stop(name, " must be a length 1 numeric.", call. = FALSE)
  }
  if (x < min){
    stop(name, " must be at least ", min, ".", call. = FALSE)
  }
  if (x > max){
    stop(name, " must be no greater than ", max, ".", call. = FALSE)
  }
  if (!round(x, 0) == x){
    stop(name, " must be a whole number.", call. = FALSE)
  }
  return()
}

#' Check if the correct starting state arguments have values
#'
#' Helper for `bmab_gi_multiple()`. Returns `TRUE` if arguments are okay, `FALSE` otherwise.
#' @noRd
check_start_args <- function(alpha_start = NULL, beta_start = NULL, Sigma_start = NULL, n_start = NULL) {
  if (is.null(alpha_start) + is.null(beta_start) == 1) return(FALSE)
  if (is.null(Sigma_start) + is.null(n_start) == 1) return(FALSE)
  return(xor(!is.null(alpha_start) && !is.null(beta_start),
             !is.null(Sigma_start) && !is.null(n_start)))
}


R/gittins-package.R
#' @description
#' This is an R package to calculate Gittins indices for the multi-armed bandit problem.
#'
#' This project contains functions written in R to calculate Gittins indices for
#' the Bayesian multi-armed bandit problem with Bernoulli or Normal rewards. Detailed information
#' on the methodology as well as the effect of different computation parameters on speed and
#' accuracy can be found in the paper at <https://arxiv.org/abs/1909.05075>
#' and in my thesis at <http://eprints.lancs.ac.uk/84589/> (in Chapter 5).
#'
#' I welcome any feedback or suggestions (see link to Github page below). If the package is useful
#'  in your work then can you please reference the package and/or the paper.
#'
#' @author James Edwards (also maintainer)
#' @seealso
#' * The main package Github page is <https://github.com/jedwards24/gittins>.
#' * Report bugs at <https://github.com/jedwards24/gittins/issues>.
#'
#' @importFrom stats dbeta dnorm integrate pbeta pnorm na.omit
#' @importFrom utils setTxtProgressBar txtProgressBar
"_PACKAGE"

## usethis namespace: start
## usethis namespace: end
NULL


R/nmab.R
#' Function arguments
#'
#' @keywords internal
#' @name nmab_args
#'
#' @param Sigma Numeric. Value of Sigma for the arm (Bayesian reward).
#' @param n Numeric > 0. Value of n for the arm (Bayesian number of observation).
#' @param gamma Numeric in (0, 1). Reward discount factor.
#' @param tau Numeric > 0. Observation precision.
#' @param tol Numeric > 0. Absolute accuracy required.
#' @param N Integer >= 2. Time horizon used.
#' @param xi Numeric > 0. Value of xi (entent of dynamic program state space).
#' @param delta Numeric > 0. Value of delta (fineness of discretisation in the dynamic program).
NULL

#' Function arguments for value functions
#'
#' @keywords internal
#' @name nmab_v_args
#'
#' @param lambda Reward from the known arm
#' @param mu Mean of reward belief for the unknown arm
#' @param n Numeric > 0. Value of n for the unknown arm
NULL

#' Calculate Gittins indices for multiple arms (normal rewards)
#'
#' This assumes mu = 0 as the GI for other values can then be derived by simple addition. See
#' `?nmab_gi` for more detail on the problem and computation parameters.
#'
#' @param n_range Numeric vector giving values of n (all greater than 0).
#' @inheritParams nmab_args
#'
#' @return A data frame of GI vales with a row for each n in `n_range`. The parameters used are
#' attached as an attribute `params`.
#'
#' @examples
#' n1 <- nmab_gi_multiple(1:4, gamma = 0.9, tau = 1, N = 30, xi = 3, delta = 0.02, tol = 5e-4)
#' n1
#' attr(n1, "params")
#' @export
nmab_gi_multiple <- function(n_range, gamma, tau, N, xi, delta, tol = 5e-4){
  if (!is.numeric(n_range)){
    stop("`n_range` must be numeric.", call. = FALSE)
  }
  if (any(n_range <= 0)){
    stop("All values in `n_range` must be strictly greater than zero.", call. = FALSE)
  }
  check_numeric(gamma, "gamma", 0, 1)

  n_range <- sort(n_range)
  nn <- length(n_range)
  gi_vec <- numeric(nn)
  ubbl <- gamma / (1 - gamma) / sqrt(n_range[1])
  ub <- nmab_giplus(0, n_range[1], gamma, tol, ub = ubbl, upper = TRUE)
  cat("Calculating GI values for", nn, "states (may be slow)\n")
  pb <- txtProgressBar(min = 0, max = nn, style = 3)
  for (i in 1:nn){
    gi_vec[i] <- nmab_gi(0, n_range[i], gamma, tau, N, xi, delta, tol = tol, ub = ub)
    ub <- gi_vec[i]
    setTxtProgressBar(pb, i)
  }
  close(pb)
  df <- data.frame(n = n_range, gi = gi_vec)
  attr(df, "params") <- list(n_range = n_range, gamma = gamma, tau = tau,
                             N = N, xi = xi, delta = delta, tol = tol)
  df
}

#' Calculate the Gittins index for a single arm (normal rewards)
#'
#' The problem state is given by the `Sigma`, `n`, `gamma`, and `tau` arguments. The
#' remaining arguments affect how the calculation is done and are chosen based on accuracy
#' and speed requirements (see details).
#'
#' @details
#' The problem has an infinite continuous state space, but the calculation can only be done for a
#' finite discrete area. `N` and `xi` control the finite extent of the state space used (one in
#' each direction). These have a diminishing effect on accuracy and just need to be sufficiently
#' large, beyond which there will be minimal improvement. Each parameter has a linear effect on the state
#' space size and therefore algorithm speed. Larger `gamma` or smaller `tau` requires
#' a larger `N`, but depends only on the least favourable of the two. Often `N` can be smaller
#' than for similar BMAB problems. The required `xi` is more robust to changes in the problem setting
#' as adjusts to `tau` somewhat. Therefore, standard values can be used across problems with less
#' disadvantage. `xi` is based on standard deviations so 3 or 4 will often be large enough. Using `N` or
#' `xi` values that are too low gives an underestimate of the Gittins index.
#'
#' The discretisation parameter `delta` is the main limiter to accuracy as it has a non-linear
#' effect on computation time. Accuracy increases as `delta` gets smaller, with inaccuracies leading
#' to an overestimate of the Gittins index.
#'
#' The `lb` and `ub` arguments can be used to provide a starting interval for calibration if desired.
#' However, for normal use this is not needed as they will be calculated internally if not supplied.
#' So the initial interval is determined as follows:
#' * For lower bound, use `lb` if supplied else use KGI.
#' * For upper bound, use `ub` if supplied else use GI+.
#'
#' @inheritParams nmab_args
#' @param lb Optional lower bound for GI.
#' @param ub Optional upper bound for GI.
#' @seealso For a link to the accompanying paper see [gittins-package].
#' @return A vector of GI values.
#'
#' @export
nmab_gi <- function(Sigma, n, gamma, tau, N, xi, delta, tol = 5e-4, lb = NA, ub = NA){
  if (is.na(lb)){
    lb <- nmab_kgi(0, n, gamma, tau, tol, ub, lower = TRUE)
  }
  if (is.na(ub)){
    ub <- nmab_giplus(0, n, gamma, tol, ub, upper = TRUE)
  }
  check_numeric(Sigma, "Sigma")
  check_numeric(n, "n", 0)
  check_numeric(gamma, "gamma", 0, 1)
  check_numeric(tau, "tau", 0)
  check_integerish(N, "N", 2L)
  check_numeric(xi, "xi", 0)
  check_numeric(delta, "delta", 0)
  check_numeric(tol, "tol", 0)
  check_numeric(lb, "lb")
  check_numeric(ub, "ub")

  interval <- Sigma / n + calibrate_arm(nmab_gi_value, lb, ub, tol, n, gamma, tau, N, xi, delta)
  mean(interval)
}

#' Calculate the GI+ index for a single arm (normal rewards)
#'
#' The index is an upper bound for GI.
#'
#' @param upper If TRUE, the upper end of the interval is returned, otherwise the midpoint.
#' @param ub Optional upper bound for the index.
#' @inheritParams nmab_args
#'
#' @return A vector of GI values.
#' @export
nmab_giplus <- function(Sigma, n, gamma, tol = 5e-4, ub = NA, upper = FALSE){
  if(is.na(ub)){ub <- gamma / (1 - gamma) / sqrt(n)}
  interval <- Sigma / n + calibrate_arm(nmab_giplus_value, lb = 0, ub = ub, tol, mu = 0, n, gamma)
  if (upper){
    return(interval[2])
  }
  mean(interval)
}

#' Calculate the knowledge gradient index for a single arm (normal rewards)
#'
#' The index is an lower bound for GI.
#'
#' @inheritParams nmab_args
#' @param lower If TRUE, the lower end of the interval is returned, otherwise the midpoint.
#' @param ub Optional upper bound for the index.
#'
#' @return A vector of GI values.
#' @export
nmab_kgi <- function(Sigma, n, gamma, tau, tol = 5e-4, ub = NA, lower = FALSE){
  if(is.na(ub)){ub <- gamma / (1 - gamma) / sqrt(n)}
  interval <- Sigma / n + calibrate_arm(nmab_kgi_value, lb = 0, ub, tol, mu = 0, n, gamma, tau)
  if (lower){
    return(interval[1])
  }
  mean(interval)
}

#' Value of one-armed bandit using GI+ (normal rewards)
#'
#' @inheritParams nmab_v_args
#' @inheritParams nmab_args
#' @keywords internal
#'
#' @return Difference in value between safe and unknown arms.
#' @export
nmab_giplus_value <- function(lambda, mu, n, gamma){
  sd <- sqrt(1 / n)
  mu + gamma * dnorm(lambda / sd) * sd / (1 - gamma * pnorm(lambda / sd)) - lambda
}

#' Value of one-armed bandit using KGI (normal rewards)
#'
#' Calculation is as given in Ryzhov, Powell & Frazier (2012), simplified since `lambda >= mu`
#' always when called here.
#'
#' @inheritParams nmab_v_args
#' @inheritParams nmab_args
#' @keywords internal
#'
#' @return Difference in value between safe and unknown arms.
#' @export
nmab_kgi_value <- function(lambda, mu, n, gamma, tau){
  sigt <- sqrt(1 / n - 1 / (n + tau))
  z <- (mu - lambda) / sigt
  v <- sigt * (z * pnorm(z) + dnorm(z))
  v * gamma / (1 - gamma) + mu - lambda
}

#' Reward of the risky arm in a one-armed bandit process
#'
#' Helper function only used in nmab_gi_value.
#' @noRd
nmab_risky_reward <- function(mu, y_lo_scaled, y_hi_scaled, tn_scaled, tau, s, value_vec, discount){
  yhi <- y_hi_scaled - mu * tn_scaled
  ylo <- y_lo_scaled - mu * tn_scaled
  p <- pnorm(yhi, mu, s) - pnorm(ylo, mu, s)
  discount * mu + (pnorm(ylo[1], mu, s) * value_vec[1] +
                     (1 - pnorm(yhi[length(yhi)], mu, s)) * value_vec[length(value_vec)] +
                     sum(p * value_vec))
}

#' Value calculation for the one-armed bandit with Normal rewards
#'
#' Assumes `Sigma = mu = 0`.
#'
#' The `extra_xi` argument was a later addition to the algorithm, not included in the paper, which
#' improves accuracy at low computational cost.
#'
#' Normally, states outside the width of the state space are ignored (taken to have a value of zero).
#' This saves computation for states that are unlikely to be visited. However, the calculation can
#' be improved with relatively little computation by giving some of these states a value using
#' their mean reward only (no further learning). Although this is an approximation it will always
#' be more accurate than using zero. So there are two blacks of states: the original states are
#' within xi standard deviations and are calculated in detail using dynamic programming; and the
#' new states within xi + extra_xi standard deviations. I have found `extra_xi = 1` works well and
#' have set this as the default. This is value that should be used unless doing research on its effect.
#'
#' @inheritParams nmab_v_args
#' @inheritParams nmab_args
#' @param extra_xi Extend xi using a fast approximation. See details
#' @keywords internal
#'
#' @return Difference in value between safe and unknown arms.
#' @export
nmab_gi_value <- function(lambda, n, gamma, tau, N, xi, delta, extra_xi = 1){
  h <- N + 1
  delta <- delta / sqrt(n) # adjust delta so number of states is constant with n
  mu_range <- seq(0, (xi + extra_xi) * sqrt(1 / n), by = delta)
  lr <- length(mu_range)
  lr2 <- length(seq(0, (xi) * sqrt(1 / n), by = delta))
  value <- matrix(nrow = h, ncol = lr)
  # Value of end states (at stage N)
  value[h, ] <- pmax(mu_range, lambda) * gamma^N / (1 - gamma)
  rr <- gamma^(0:N) / (1 - gamma)
  value[, (lr2:lr)] <- matrix(rr, ncol = 1) %*% matrix(mu_range[lr2:lr], nrow = 1)
  lo <- mu_range - delta / 2 # lower end of discrete approximation to mu
  hi <- mu_range + delta / 2
  for (j in N:2){
    t <- j - 1
    tn <- n + tau * (j - 1)
    #the next 3 variables are used for speed-up only
    y_hi_scaled <- hi * (tn + tau) / tau
    y_lo_scaled <- lo * (tn + tau) / tau
    tn_scaled <- tn / tau
    s <- sqrt(1 / tn + 1 / tau) #sd of y
    discount <- gamma^t
    safe_reward <- lambda * discount / ( 1 - gamma)
    value_vec <- value[j + 1, ]
    for (i in (lr2 - 1):1){
      risky_reward <- nmab_risky_reward(mu_range[i], y_lo_scaled, y_hi_scaled, tn_scaled,
                                        tau, s, value_vec, discount)
      if (risky_reward > safe_reward){
        value[j, i] <- risky_reward
      }else{
        value[j, 1:i] <- safe_reward
        break
      }
    }
    # If risky arm preferred in [j, 1] then it will be preferred in starting state
    if (value[j, 1] > safe_reward){
      return(value[j, 1] - safe_reward)
    }
  }
  # Value of risky arm in starting state at time 0
  s <- sqrt(1 / n + 1 / tau)
  value_vec = value[2, ]
  risky_reward <- nmab_risky_reward(mu_range[1],
                                    y_lo_scaled = lo * (n + tau) / tau,
                                    y_hi_scaled = hi * (n + tau) / tau,
                                    tn_scaled = n / tau,
                                    tau, s, value_vec, discount = 1)
  return(risky_reward - lambda / (1 - gamma))
}
